lua/metals/rootdir.lua

```diff
@@ -11,32 +11,42 @@ local has_pattern = function(patterns, target)
   end
 end
 
---- NOTE: This only searches 2 levels deep to find nested build files.
+--- NOTE: maxParentSearch lets you check up to a certain number of parent
+--- folders to find nested build files.
 --- Given a situation like the below one where you have a root build.sbt
 --- and one in your module a, you want to ensure the root is correctly set as
 --- the root one, not the a one. This checks the parent dir to ensure this.
 --- build.sbt  <-- this is the root
 --- a/
 ---  - build.sbt <- this is not
 ---  - src/main/scala/Main.scala
-local find_root_dir = function(patterns, startpath)
+--- If your projects are multiple layers deep, set
+--- config.find_root_dir_max_project_nesting to a greater number. Default is 1
+--- for the behavior described above.
+local find_root_dir = function(patterns, startpath, maxParentSearch)
   local path = Path:new(startpath)
   -- TODO if we don't find it do we really want to search / probably not... add a check for this
-  for _, parent in ipairs(path:parents()) do
+  -- First parent index in which we found a target file
+  local firstFoundIdx = -1
+  local ret = nil
+
+  for i, parent in ipairs(path:parents()) do
+    -- Exit loop before checking anything if we've exceeded the search limits
+    if firstFoundIdx >= 0 and (i - firstFoundIdx > maxParentSearch) then
```

Comment:
Shouldn't this be

```suggestion
    if firstFoundIdx >= 0 and (i - firstFoundIdx >= maxParentSearch) then
```

For example if you open up a project like:

```
.scala-build/
.bsp/
project.scala
```

Which is a valid scala-cli workspace, and you don't have `maxParentSearch` set so it defaults to 1, with this being `>` you end up searching 2 layers down, when you only want to search one.

Reply:
Actually to further simplify this, wouldn't this also work?

```lua
local find_root_dir = function(patterns, startpath, maxParentSearch)
  local path = Path:new(startpath)
  -- TODO if we don't find it do we really want to search / probably not... add a check for this
  -- First parent index in which we found a target file
  local firstFoundIdx = nil
  local ret = nil

  for i, parent in ipairs(path:parents()) do
    -- Exit loop before checking anything if we've exceeded the search limits
    if firstFoundIdx and (i > maxParentSearch) then
      return ret
    end
    local pattern = has_pattern(patterns, parent)
    if pattern then
      -- Mark the first parent that was found, so we can exit the loop when we've exhausted our search limits
      if not firstFoundIdx then
        firstFoundIdx = i
      end
      -- (over)write the return value with the highest parent found
      ret = parent
    end
  end
  -- In case we went through the entire loop (e.g. if maxParentSearch is really high)
  return ret
end
```

Reply:
Re the first comment: The way I interpreted the setting is that the default behavior (which is nvim-metals' behavior prior to this change) searches *1 parent folder up*, and if build.sbt is in the same folder as project.scala then that's *0 parent folders up* because it's the same folder.
Re the second comment and code suggestion: I don't have access currently to the server that I was developing on, so I can't test it at the moment. But I trust your Lua skills, so if it's got the same behavior I'm cool with it!